src/realm/object-store/schema.cpp

```diff
@@ -348,20 +359,39 @@ std::vector<SchemaChange> Schema::compare(Schema const& target_schema, SchemaMod
     return changes;
 }
 
-void Schema::copy_keys_from(realm::Schema const& other) noexcept
+void Schema::copy_keys_from(realm::Schema const& other, bool is_schema_additive) noexcept
 {
-    zip_matching(*this, other, [&](ObjectSchema* existing, const ObjectSchema* other) {
-        if (!existing || !other)
-            return;
-
-        existing->table_key = other->table_key;
-        for (auto& current_prop : other->persisted_properties) {
-            auto target_prop = existing->property_for_name(current_prop.name);
-            if (target_prop) {
-                target_prop->column_key = current_prop.column_key;
+    auto other_classes = zip_matching(
+        *this, other,
+        [&](ObjectSchema* existing, const ObjectSchema* other) {
+            if (!existing || !other)
+                return;
+
+            std::vector<Property> unmatched_properties;
+            existing->table_key = other->table_key;
+            for (auto& current_prop : other->persisted_properties) {
+                auto target_prop = existing->property_for_name(current_prop.name);
+                if (target_prop) {
+                    target_prop->column_key = current_prop.column_key;
+                }
+                else if (is_schema_additive) {
+                    unmatched_properties.push_back(current_prop);
+                }
             }
-        }
-    });
+
+            if (is_schema_additive) {
+                existing->persisted_properties.insert(existing->persisted_properties.end(),
+                                                      unmatched_properties.begin(), unmatched_properties.end());
+            }
+        },
+        is_schema_additive);
+
+    if (!other_classes.empty()) {
+        insert(end(), other_classes.begin(), other_classes.end());
+        std::sort(begin(), end(), [](ObjectSchema const& lft, ObjectSchema const& rgt) {
```

Comment:
can you extract this to method just to be sure that we are sorting things the same way wherever else we sort? (the one other place I know of is the Schema constructor)